### PR TITLE
:new: Define `DeepFeatureExtractor`

### PR DESCRIPTION
**Pull request by shaneahmed**
# 🚀 Summary

This PR introduces a new **`DeepFeatureExtractor` engine** to the TIAToolbox framework, enabling extraction of intermediate CNN feature representations from whole slide images (WSIs) or image patches. These features can be used for downstream tasks such as clustering, visualization, or training other models. The update also includes:

- A **command-line interface (CLI)** for the new engine.
- Extended **CLI utilities** for flexible input/output configurations.
- Comprehensive **unit tests** covering patch-based and WSI-based workflows, multi-GPU support, and CLI functionality.
- Integration with TIAToolbox’s model registry and CLI ecosystem.

---

### **New Engine: `DeepFeatureExtractor`**
- Extracts intermediate CNN features from WSIs or patches.
- Outputs feature embeddings and spatial coordinates in **Zarr** or **dict** format.
- Implements **memory-aware caching** for large-scale WSI processing.
- Compatible with:
  - TIAToolbox pretrained models.
  - Torchvision CNN backbones (e.g., ResNet, DenseNet, MobileNet).
  - **All timm architectures via `timm.list_models()`**, including HuggingFace-hosted models.
- Supports both **patch-mode** and **WSI-mode** workflows.

### **CLI Integration**
- Adds `deep-feature-extractor` command to TIAToolbox CLI.
- Supports options for:
  - Input/output paths and file types.
  - Model selection (`resnet18`, `efficientnet_b0`, timm-based backbones, etc.).
  - Patch extraction parameters (`patch_input_shape`, `stride_shape`, `input_resolutions`).
  - Batch size, device selection, memory threshold, overwrite behavior.
- Flexible JSON-based CLI options for resolutions and class mappings.

### **Extended CLI Utilities**
- New reusable options:
  - `--input-resolutions`, `--output-resolutions` (JSON list of dicts).
  - `--patch-input-shape`, `--stride-shape`, `--scale-factor`.
  - `--class-dict` for mapping class indices to names.
  - `--overwrite` and `--output-file` for fine-grained control.

### **Unit Tests**
- **Engine Tests**:
  - Patch-based and WSI-based feature extraction.
  - Validation of Zarr outputs (features and coordinates).
  - Multi-GPU functionality.
- **Model Compatibility**:
  - Tests with `CNNBackbone` and `TimmBackbone` models.
- **CLI Tests**:
  - Single-file and parameterized runs.
  - Validation of JSON parsing for CLI options.

### **Codebase Integration**
- Registers `DeepFeatureExtractor` in `tiatoolbox.models` and engine registry.
- Adds CLI command in `tiatoolbox.cli.__init__.py`.
- Updates architecture utilities to support timm-based backbones and HuggingFace models.
- Introduces dictionaries for Torch and timm backbones (`torch_cnn_backbone_dict`, `timm_arch_dict`).

### CODE REVIEW

File: tests/engines/test_feature_extractor.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    patches = patches.type(torch.float32)
+    model = model.to("cpu")
+    # Inference mode
+    model.eval()
+    with torch.inference_mode():
+        _features = model(patches).numpy()
+    # ! must maintain same batch size and likely same ordering
+    # ! else the output values will not exactly be the same (still < 1.0e-4
+    # ! of epsilon though)
+    assert np.mean(np.abs(features[:4] - _features)) < 1.0e-1
```
Comment (shaneahmed) [inline tests/engines/test_feature_extractor.py:160]: For some reason, the error is quite high. Not clear about the issue yet. I will update once I have resolved the issue.

File: tiatoolbox/models/engine/deep_feature_extractor.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        self: SemanticSegmentor,
+        dataloader: DataLoader,
+        save_path: Path,
+        **kwargs: Unpack[SemanticSegmentorRunParams],
+    ) -> dict[str, da.Array]:
+        """Perform model inference on a whole slide image (WSI).
+
+        This method processes a WSI using the provided DataLoader, merges
+        patch-level predictions into a full-resolution canvas, and returns
+        the aggregated output. It supports memory-aware caching and optional
```
Comment (measty) [inline tiatoolbox/models/engine/deep_feature_extractor.py]: this docstring is still holdover from the semanticSegemntor; we aren't making any canvas here right?

File: tiatoolbox/models/engine/deep_feature_extractor.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+            probabilities.append(da.from_array(batch_output[0]))
+            coordinates.append(
+                da.from_array(
+                    self._get_coordinates(batch_data),
+                )
+            )
+
+        raw_predictions["coordinates"] = da.concatenate(coordinates, axis=0)
+        raw_predictions["probabilities"] = da.concatenate(probabilities, axis=0)
```
Comment (measty) [inline tiatoolbox/models/engine/deep_feature_extractor.py]: various references to probabilities throughout this function, but feature extractor isnt doing anything with probabilities so it's a bit of a misleading name.

Either change it to something more appropriate in here, or maybe change it in the base SemanticSegmentor to something more generic?

File: tiatoolbox/models/engine/deep_feature_extractor.py [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+            dict[str, dask.array.Array]:
+                Dictionary containing:
+                - "probabilities": Extracted feature maps from the model.
+                - "coordinates": Patch coordinates corresponding to the features.
+
+        """
+        # Default Memory threshold percentage is 80.
+        memory_threshold = kwargs.get("memory_threshold", 80)
+        vm = psutil.virtual_memory()
+        keys = ["probabilities", "coordinates"]
```
Comment (shaneahmed) [inline tiatoolbox/models/engine/deep_feature_extractor.py:286]:
```suggestion
        keys = ["features", "coordinates"]
```

File: tiatoolbox/models/engine/deep_feature_extractor.py [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+                Path to model weights. If None, default weights are used.
+            device (str):
+                Device to run the model on (e.g., "cpu", "cuda"). Default is "cpu".
+            verbose (bool):
+                Whether to enable verbose logging. Default is True.
+
+        """
+        super().__init__(
+            model=model,
+            batch_size=batch_size,
```
Comment (measty) [inline tiatoolbox/models/engine/deep_feature_extractor.py:213]: when a string is passed as a model, it should really go look in the lists of architectures available for TIMMBackbone and CNNBackbone (in architecture.vanilla.py) and build the appropriate model. At the moment it looks in the models defined in pretrainedModel.yaml which is patch prediction/segmentation models. And those don't work in here.

We may even want to consolidate these, so that there is a single unified way of getting models, whatever type they are.

File: tiatoolbox/models/engine/deep_feature_extractor.py [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+        images: list[os.PathLike | Path | WSIReader] | np.ndarray,
+        masks: list[os.PathLike | Path] | np.ndarray | None = None,
+        labels: list | None = None,
+        ioconfig: IOSegmentorConfig | None = None,
+        *,
+        patch_mode: bool = True,
+        save_dir: os.PathLike | Path | None = None,
+        overwrite: bool = False,
+        output_type: str = "dict",
+        **kwargs: Unpack[SemanticSegmentorRunParams],
```
Comment (measty) [inline tiatoolbox/models/engine/deep_feature_extractor.py]: This has some usability issues. As a user (and especially trying to put myself into the shoes of someone relatively new to tiatoolbox) I would just want to specify a model, a patch size, a resolution, some data to run it on, and somewhere to save. It should be intuitive how to provide those.

In practice, some of those things are hidden away - i either have to go off into IOSegmentorConfig and figure out how to make one of those, or hunt around in SemanticSegmentorRunParams to find out from these how to specify them

File: tests/engines/test_feature_extractor.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+# Command Line Interface
+# -------------------------------------------------------------------------------------
+
+
+def test_cli_model_single_file(sample_svs: Path, track_tmp_path: Path) -> None:
+    """Test for feature extractor CLI single file."""
+    runner = CliRunner()
+    models_wsi_result = runner.invoke(
+        cli.main,
+        [
```
Comment (measty) [inline tests/engines/test_feature_extractor.py:215]: This is not a correct test.

what this is actually doing:
As default model is fcn-tissue-mask, it is running that within the DeepFeatureExtractor engine and saving the patches that come out of that (the probability maps for tissue/non-tissue) in a wierd zarr array that has shape (512*num_patches, 512, 2)

In practice, the deep-feature-extractor cli isnt useable because of the issue i pointed out in one of my other comments with providing model as string; you cant pass name of a valid CNNBackbone or TimmBackbone in there.

File: tiatoolbox/models/engine/deep_feature_extractor.py [hunk 8]
```diff
@@ -0,0 +1,557 @@
+"""Deep Feature Extraction Engine for Digital Pathology.
+
+This module defines the `DeepFeatureExtractor` class, which extends
+`SemanticSegmentor` to extract intermediate CNN feature representations
```
Comment (measty) [inline tiatoolbox/models/engine/deep_feature_extractor.py]: It isn't just CNN features (most foundation models are transformer)

File: tiatoolbox/models/engine/deep_feature_extractor.py [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+            *coordinates_zarr.shape[1:],
+        )
+    )
+    coordinates_zarr[-coordinates_computed.shape[0] :] = coordinates_computed
+
+    return probabilities_zarr, coordinates_zarr
+
+
+class DeepFeatureExtractor(SemanticSegmentor):
+    r"""Generic CNN-based feature extractor for digital pathology images.
```
Comment (measty) [inline tiatoolbox/models/engine/deep_feature_extractor.py]: remove CNN-based

File: tiatoolbox/models/engine/deep_feature_extractor.py [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+---------
+- Handles large-scale WSIs with memory-aware caching.
+- Outputs feature maps and patch coordinates for downstream analysis.
+- Compatible with TIAToolbox pretrained models and custom PyTorch models.
+- Supports both patch-based and WSI-based workflows.
+
+Example:
+--------
+>>> from tiatoolbox.models.engine.deep_feature_extractor import DeepFeatureExtractor
+>>> extractor = DeepFeatureExtractor(model="resnet50-kather100k")
```
Comment (measty) [inline tiatoolbox/models/engine/deep_feature_extractor.py]: This example wouldn't work as resnet50-kather100k isnt a feature extraction model (and wouldnt run as patch_output_shape isnt defined in it's ioconfig)

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 8 above)
Comment (shaneahmed) [inline tiatoolbox/models/engine/deep_feature_extractor.py]:
```suggestion
`SemanticSegmentor` to extract intermediate feature representations
```

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 9 above)
Comment (shaneahmed) [inline tiatoolbox/models/engine/deep_feature_extractor.py]:
```suggestion
    r"""Generic deep feature extractor for digital pathology images.
```

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 5 above)
Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py:213]: Ideally, the feature extractor should work with both `TimmBackbone("efficientnet_b0", pretrained=True)` and `"efficientnet_b0"` options.

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 6 above)
Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py]: Explicitly specify input_resolutions and patch_input_shape instead of wrapping it into SemanticSegmentorRunParams.

File: tiatoolbox/models/engine/deep_feature_extractor.py [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+            patch_mode (bool):
+                Whether to treat input as patches (`True`) or WSIs (`False`).
+                Default is True.
+            save_dir (PathLike | None):
+                Directory to save output files. Required for WSI mode.
+            overwrite (bool):
+                Whether to overwrite existing output files. Default is False.
+            output_type (str):
+                Desired output format. Must be "zarr" or "dict".
+            **kwargs (SemanticSegmentorRunParams):
```
Comment (shaneahmed) [inline tiatoolbox/models/engine/deep_feature_extractor.py]: Expand the docstring for all the options.

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 8 above)
Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py]: Resolved in https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/1080d5db0ba9250a9615d0186f352c46db3f64f5

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 9 above)
Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py]: Resolved in https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/1080d5db0ba9250a9615d0186f352c46db3f64f5

Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py]: Resolved in https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/1080d5db0ba9250a9615d0186f352c46db3f64f5

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 8 above)
Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py]: Resolved in https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/1080d5db0ba9250a9615d0186f352c46db3f64f5

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 11 above)
Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/98e2312b2cb1aa133931485d7293a13b21672908

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 6 above)
Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py]: Addressed expanded docstrings in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/98e2312b2cb1aa133931485d7293a13b21672908

Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py]: `input_resolutions` and `patch_input_shape` explicitly specified in https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/ab80306ad4a2a0bdc36d091c08f0bd74dd14de34 and https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/305588fdae3f7e667cb630bbe89870b30b82532e

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 3 above)
Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py]: Renaming intermediate output will mean updating all the base classes to use new keys. I rename this after processing to "features" in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/6c3b82106938e0042cee8e830acc47b4c890af69

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 5 above)
Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py:213]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/d24dfbe0081b4fe842048a22a80befd74f8ce141

File: tiatoolbox/models/engine/deep_feature_extractor.py (hunk 10 above)
Comment (shaneahmed) [reply tiatoolbox/models/engine/deep_feature_extractor.py]: Feature Extractor can be based on PatchPredictor. This should work with any timm model. I have updated it in this commit 
https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/d24dfbe0081b4fe842048a22a80befd74f8ce141
and https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/cce3883d4af4f27d6597e9e38bad157c9fa988dd

File: tests/engines/test_feature_extractor.py (hunk 7 above)
Comment (shaneahmed) [reply tests/engines/test_feature_extractor.py:215]: Addressed it in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/c508549c4b3107bc88b111d13296c5e86ec0ff4c

CLI now supports all the inputs except labels. 

The expected output shape should be `# of patches x feature size`

File: tests/engines/test_feature_extractor.py [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+    # ! else the output values will not exactly be the same (still < 1.0e-4
+    # ! of epsilon though)
+    assert np.mean(np.abs(features[:4] - _features)) < 1.0e-1
+
+
+@pytest.mark.skipif(
+    toolbox_env.running_on_ci() or not ON_GPU,
+    reason="Local test on machine with GPU.",
+)
+def test_multi_gpu_feature_extraction(
```
Comment (shaneahmed) [inline tests/engines/test_feature_extractor.py:167]: @adamshephard Please can you test this? Thanks

File: tiatoolbox/models/architecture/__init__.py [hunk 13]
```diff
@@ -127,6 +130,12 @@ def get_pretrained_model(
         msg = "pretrained_model must be a string."
         raise TypeError(msg)
 
+    if pretrained_model in torch_cnn_backbone_dict:
+        return CNNBackbone(pretrained_model), None
+
```
Comment (measty) [inline tiatoolbox/models/architecture/__init__.py:135]: needs a block here to look in timm_arch_dict from vanilla.py

Otherwise something like:
 extractor = DeepFeatureExtractor(model="UNI")
 will not work

**Review by measty — CHANGES_REQUESTED** [review]
Looks much better now, just one last thing to address and then i'm happy to approve

File: tiatoolbox/models/architecture/__init__.py (hunk 13 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/__init__.py:135]: Thanks. It is now addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/963/commits/82ab2c4d59651206ebd76d036b1fae687963a3dd

We will need to wait for Adam to pass the local test for UNI before we merge it.
